kb/2023-08-11/using_netapp_third_party_storage.md

```diff
@@ -0,0 +1,281 @@
+
+   1. Put the following into a text file, for example /tmp/backend.yaml
+
+      ```yaml
+      version: 1
+      backendName: default_backend_san
+      storageDriverName: ontap-san-economy
+      managementLIF: 172.19.97.114
+      svm: default_backend
+      username: admin
+      password: password1234
+      labels:
+      name: default_backend_san
+      ```
+
+    The LIF IP address, username, and password of this file
+    should be replaced with the management LIF and credentials
+    for the ONTAP system.
+
+   1. Create the backend
+
+      ```shell
+      ./tridentctl create backend -f /tmp/backend.yaml -n trident
+      ```
+
+   1. Check that it is created
+
+      ```shell
+      ./tridentctl get backend -n trident
+      ```
+
+1. Define a StorageClass and SnapshotClass.
+
+   1. Put the following into a file, for example `/tmp/storage.yaml`
+
+      ```yaml
+      ---
+      apiVersion: storage.k8s.io/v1
+      kind: StorageClass
+      metadata:
+      name: ontap-san-economy
+      provisioner: csi.trident.netapp.io
+      parameters:
+      selector: "name=default_backend_san"
+      ---
+      apiVersion: snapshot.storage.k8s.io/v1
+      kind: VolumeSnapshotClass
+      metadata:
+      name: csi-snapclass
+      driver: csi.trident.netapp.io
+      deletionPolicy: Delete
+      ```
+
+   1. Apply the definitions:
+
+      ```shell
+      kubectl apply -f /tmp/storage.yaml
+      ```
+
+1. Enable multipathd
+
+   The following is required to enable multipathd.
+   This must be done on every node of the Harvester cluster.
+   The preceding steps should only be done once on a single node.
+
+   1. Create this file in `/oem/99_multipathd.yaml`:
+
+      ```yaml
+      stages:
+         default:
+         - name: "Setup multipathd"
+            systemctl:
+               enable:
+               - multipathd
+               start:
+               - multipathd
+      ```
+
+   1. Configure `multipathd` to exclude pathnames used by Longhorn.
+
+      This part is a little tricky.  `multipathd` will automatically discover
+      device names matching a certain pattern, and attempt to set up multipathing on them.
+      Unfortunately, Longhorn's device names follow the same pattern, and
+      will not work correctly if `multipathd` tries to use those devices.
+
+      Therefore the file `/etc/multipath.conf` must be set up on each node
+      so as to prevent `multipathd` from touching any of the devices
+      that Longhorn will use.  Unfortunately, it is not possible to know
+      in advance which device names will be used until the volumes are attached
+      to a VM when the VM is started, or when the volumes are hot-added to a running VM.
+      The recommended method is to "whitelist" the Trident devices using device
+      properties rather than device naming.  The properties to allow are the
+      device vendor and product.  Here is an example of what you'll want in `/etc/multipath.conf`:
```

Comment:
```suggestion
   1. Configure `multipathd` to exclude pathnames used by Longhorn.

      This part is a little tricky. `multipathd` will automatically discover device names matching a specific pattern and attempt to set up multipathing on them. Unfortunately, Longhorn's device names follow the same pattern and will not work correctly if `multipathd` tries to use those devices.

      Therefore, the file `/etc/multipath.conf` must be set up on each node to prevent `multipathd` from touching any devices that Longhorn will use. Unfortunately, it is not possible to know in advance which device names will be used until the volumes are attached to a VM when the VM is started or when the volumes are hot-added to a running VM. The recommended method is to "whitelist" the Trident devices using device properties rather than device naming. The properties to allow are the device vendor and product. Here is an example of what you'll want in `/etc/multipath.conf`:
```